includes/class-omise-callback.php

```diff
@@ -0,0 +1,178 @@
+<?php
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * @since 4.0
+ */
+class Omise_Callback {
```

Comment:
Three methods in here take both an order and a charge. That's refactoring 101 for "here's your state".

So, `execute` should get the order, make a `new Callback` set the order and finally call a member function (`execute` like the static is fine I guess). This function finds the appropriate charge, then `switch` on the `$charge['status']` and decide which private function to call (because there's no point in making those public).

Benefit is that now, if you need to run this manually, you can skip the static execute (which uses `$_GET['order_id']` to find the order), and instead instantiate a callback, set the order and call `->execute()`.

Reply:
@robinclart Been thinking about this, and I like the idea of manually set order so we can skip the static execute if needed. But probably at this moment, where there is still no case (and never has a potential) that would need that approach, how about we just keep the code as simple as it is and let the code gets evolved when it's needed?
What do you think?

Reply:
Did you actually try this out and judged which one looked/read better?

Reply:
To me the code doesn't look simple as it is. That $order, $charge tuple you're passing around screams "refactor me". And the whitelist still looks more complicated than it should. I just would not do the dynamic function calling. The added benefit is that you can pick better names for your methods as well if you wish.

Reply:
@robinclart was thinking and designing a bit but not completely finish the code
However, just for context, originally, I decided to come for this approach as I want to leave the code programatically hook-able for any 3rd-party developers that they want to alter the "result" of `return_uri` callback behaviour. 
i.e. in WordPress, developers often use `add_action()` or `add_filter()` functions to hook and alter a specific point of code. 
The same as the way we are able to hook our callback function to WC as in the following code
```php
add_action( 'woocommerce_api_omise_alipay_callback', 'Omise_Callback::execute' );
```

That, is also the reason why I was deciding to pass `$order` and `$charge` as an argument instead of class properties, so, in the future, if we want to open these methods to be able to be hooked, then we can update the core code a bit, and developers can simply implement, on their theme directory:
```php
// function.php
add_action( 'omise_callback_payment_successful', 'my_payment_successful', 0, 2 );
function my_payment_successful( $order, $charge ) {
    // ... then they can alter from this point to serve their desire website flow
}
```

Or even as for extending Omise_Callback class. As, it'd be easier that developers don't need to look up what properties the class has, but do whatever in a scope of method's arguments.

--
But, give me some time to design the code for your suggestion a bit. I will code something out.

Reply:
I don't think what you're saying and what I'm saying is incompatible.